src/tools/rbd_mirror/ImageMap.cc

```diff
@@ -225,6 +238,65 @@ void ImageMap<I>::schedule_update_task() {
   m_threads->timer->add_event_after(after, m_timer_task);
 }
 
+template <typename I>
+void ImageMap<I>::rebalance(double timeout) {
+  assert(timeout != 0);
+  assert(m_rebalance_task == nullptr);
+
+  {
+    Mutex::Locker locker(m_lock);
+    if (m_async_op_tracker.empty() &&
+        m_global_image_ids.empty() &&
+        ((ceph_clock_now() - m_last_policy_update_time) >= timeout)) {
+
+      dout(20) << "starting rebalance" << dendl;
+
+      std::set<std::string> remap_global_image_ids;
+      m_policy->add_instances({}, &remap_global_image_ids);
+
+      for (auto const &global_image_id : remap_global_image_ids) {
+        schedule_action(global_image_id);
+      }
+
+      // set policy update time if there is nothing to process, so
+      // that we do not reenter here again
+      if (m_global_image_ids.empty()) {
+        m_last_policy_update_time = ceph_clock_now();
+      }
+    }
+  }
+
+  schedule_update_task(m_threads->timer_lock);
+}
+
+template <typename I>
+void ImageMap<I>::schedule_rebalance_task() {
+  assert(m_threads->timer_lock.is_locked());
+
+  // fetch the updated value of idle timeout for (re)scheduling
+  CephContext *cct = reinterpret_cast<CephContext *>(m_ioctx.cct());
+  double resched_after =
+    cct->_conf->get_val<double>("rbd_mirror_image_policy_rebalance_on_idle_timeout");
+
+  if (!resched_after || m_rebalance_task != nullptr) {
```

Comment:
Nit: you could just cancel any previously scheduled event re-schedule a new one. Then you could drop the ```m_last_policy_update_time``` logic.

Reply:
Wouldn't that be too many cancellations and reschedules as `schedule_rebalance_task()` gets called from `schedule_update_task()`?

Reply:
I wouldn't think it would be a CPU performance concern. Alternatively, you could call   ```schedule_rebalance_task``` only if a new update task was scheduled (i.e. every second).

Reply:
OK -- I kept it simple as of now and invoke `schedule_rebalance_task()` early on from `schedule_update_task()`